pkg/reconciler/v1alpha1/revision/table_test.go

```diff
@@ -588,6 +588,90 @@ func TestReconcile(t *testing.T) {
 				WithSucceededFalse("SomeReason", "This is why the build failed.")),
 		},
 		Key: "foo/failed-build-stable",
+	}, {
+		Name: "ready steady state",
+		// Test the transition that Reconcile makes when Endpoints become ready.
+		// This puts the world into the stable post-reconcile state for an Active
+		// Revision.  It then creates an Endpoints resource with active subsets.
+		// This signal should make our Reconcile mark the Revision as Ready.
+		Objects: []runtime.Object{
+			rev("foo", "steady-ready", WithK8sServiceName, WithLogURL,
+				// When the endpoint and KPA are ready, then we will see the
+				// Revision become ready.
+				MarkRevisionReady),
+			kpa("foo", "steady-ready", WithTraffic),
+			deploy("foo", "steady-ready"),
+			svc("foo", "steady-ready"),
+			endpoints("foo", "steady-ready", WithSubsets),
+			image("foo", "steady-ready"),
+		},
+		WantEvents: []string{
+			Eventf(corev1.EventTypeNormal, "RevisionReady", "Revision becomes ready upon endpoint %q becoming ready",
+				"steady-ready-service"),
+		},
+		Key: "foo/steady-ready",
+	}, {
+		Name:    "lost service owner ref",
+		WantErr: true,
+		Objects: []runtime.Object{
+			rev("foo", "missing-owners", WithK8sServiceName, WithLogURL,
+				MarkRevisionReady),
+			kpa("foo", "missing-owners", WithTraffic),
+			deploy("foo", "missing-owners"),
+			svc("foo", "missing-owners", WithK8sSvcOwnersRemoved),
+			endpoints("foo", "missing-owners", WithSubsets),
+			image("foo", "missing-owners"),
+		},
+		WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+			Object: rev("foo", "missing-owners", WithK8sServiceName, WithLogURL,
+				MarkRevisionReady,
+				// When we're missing the OwnerRef for Service we see this update.
+				MarkResourceNotOwned("Service", "missing-owners-service")),
+		}},
+		Key: "foo/missing-owners",
+	}, {
+		Name:    "lost kpa owner ref",
+		WantErr: true,
+		Objects: []runtime.Object{
+			rev("foo", "missing-owners", WithK8sServiceName, WithLogURL,
+				MarkRevisionReady),
+			kpa("foo", "missing-owners", WithTraffic, WithPodAutoscalerOwnersRemoved),
+			deploy("foo", "missing-owners"),
+			svc("foo", "missing-owners"),
+			endpoints("foo", "missing-owners", WithSubsets),
+			image("foo", "missing-owners"),
+		},
+		WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+			Object: rev("foo", "missing-owners", WithK8sServiceName, WithLogURL,
+				MarkRevisionReady,
+				// When we're missing the OwnerRef for PodAutoscaler we see this update.
+				MarkResourceNotOwned("PodAutoscaler", "missing-owners")),
+		}},
+		// TODO(mattmoor): This seems strange.
```

Comment:
It does indeed. I think it's currently hardcoded that once endpoints become ready, everything is "fine". We could move the event generation out of the `reconcile_resources.go` file towards the top-level `revision.go` file and produce it only if the incoming revision was not ready, but the reconciled revision is?

Potentially, that'd also fix your comment close to the event generation saying:

```
// TODO(mattmoor): How to ensure this only fires once?
```

Reply:
Yeah, I agree.  I left this breadcrumb and forgot to circle back, so thanks for keeping me honest! 🎉 

As this is fairly orthogonal to this change, I'm going to open an issue to track this and change the `TODO(#1234)` to reference it.